lib/websocket/frame.js

```diff
@@ -0,0 +1,223 @@
+'use strict'
+
+const { randomBytes } = require('crypto')
+const assert = require('assert')
+const { opcodes, maxUnsigned16Bit } = require('./constants')
+const { isValidStatusCode } = require('./util')
+
+class WebsocketFrame {
+  #offset = 0
+  #buffers = []
+  #fragmentComplete = false
+
+  /**
+   * Whether a frame (unfragmented or fragmented) is complete.
+   */
+  get terminated () {
+    return this.#fragmentComplete
+  }
+
+  set terminated (value) {
+    this.#fragmentComplete = value
+  }
+
+  /*
+  https://www.rfc-editor.org/rfc/rfc6455#section-5.2
+  0                   1                   2                   3
+  0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
+ +-+-+-+-+-------+-+-------------+-------------------------------+
+ |F|R|R|R| opcode|M| Payload len |    Extended payload length    |
+ |I|S|S|S|  (4)  |A|     (7)     |             (16/64)           |
+ |N|V|V|V|       |S|             |   (if payload len==126/127)   |
+ | |1|2|3|       |K|             |                               |
+ +-+-+-+-+-------+-+-------------+ - - - - - - - - - - - - - - - +
+ |     Extended payload length continued, if payload len == 127  |
+ + - - - - - - - - - - - - - - - +-------------------------------+
+ |                               |Masking-key, if MASK set to 1  |
+ +-------------------------------+-------------------------------+
+ | Masking-key (continued)       |          Payload Data         |
+ +-------------------------------- - - - - - - - - - - - - - - - +
+ :                     Payload Data continued ...                :
+ + - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - +
+ |                     Payload Data continued ...                |
+ +---------------------------------------------------------------+
+ */
+  constructor ({
+    opcode = opcodes.TEXT,
+    fin = false,
+    rsv1 = false,
+    rsv2 = false,
+    rsv3 = false,
+    payloadLength,
+    data
+  } = {}) {
+    this.fin = fin
+    this.rsv1 = rsv1
+    this.rsv2 = rsv2
+    this.rsv3 = rsv3
+    this.opcode = opcode
+
+    this.#buffers = [data]
+    this.#offset += data.length
+
+    this.payloadLength = payloadLength
+    this.fragmented = !this.fin && this.opcode !== opcodes.CONTINUATION
+  }
+
+  get data () {
+    return Buffer.concat(this.#buffers, this.#offset)
```

Comment:
if buffers.length === 1 we don't need to make a copy? Also should we memoize this once it is done?

Reply:
I've ensured that this getter is never used more than once per code path, I'll add a fast path for 1 buffer.